### PR TITLE
Poi locations

### PR DESCRIPTION
**Pull request by jim-martin**
**Description of changes**

Adds a Location Prefab tab to the AbstractMap UI, and convenience methods in AbstractMap. Features to look at:
1. Location Prefabs section of `AbstractMap`
2. Placing prefabs using categories
   - Location Prefabs > Findby `Mapbox Category`
3. Placing prefabs using place names
   - Location Prefabs > Findby `Name`
4. Placing prefabs using lat longs
   - Location Prefabs > Findby `Address or Lat Lon`
5. Searching for locations (ui components with the `Geocode` tag)
   - Any field tagged with the `[Geocode]` attribute
6. The `LocationPrefabs.unity` example scene
7. **All** Scenes using Prefab Modifiers
8. scripts implementing `IFeaturePropertySettable`
9. scripts using `LocationPrefabCategoryOptions`
10. API functionality - `LocationPrefabsScriptExample.cs`

**Known Issues**
1. Features are placed on tile center, but prefabs are placed at the correct feature position.

**QA checklists**

- [x] Add relevant code comments. Every API class and method should have `<summary>` description as well as description of parameters.
- [ ] **Add tests for new/changed/updated classes and methods!!!**
- [x] Check out conventions in [CONTRIBUTING.md](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/CONTRIBUTING.md).
- [x] Check out conventions in [CODING-STYLE.md](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/CODING-STYLE.md)
- [x] Update the [changelog](https://github.com/mapbox/mapbox-unity-sdk/blob/develop/documentation/docs/05-changelog.md)
- [x] Update documentation.

**Reviewers**

Tag your reviewer(s). Choose wisely.

### CODE REVIEW

File: sdkproject/Assets/Mapbox/Unity/Editor/GeocodeAttributeSearchWindow.cs [hunk 1]
```diff
@@ -83,15 +84,44 @@ void OnGUI()
...
 						if (GUILayout.Button(buttonContent))
 						{
-							_property.stringValue = coordinates;
+							_coordinateProperty.stringValue = coordinates;
+							_coordinateProperty.serializedObject.ApplyModifiedProperties();
+							EditorUtility.SetDirty(_coordinateProperty.serializedObject.targetObject);
 
-							_property.serializedObject.ApplyModifiedProperties();
-							EditorUtility.SetDirty(_property.serializedObject.targetObject);
+							//TODO: assign the address string to an ajacent property.
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Editor/GeocodeAttributeSearchWindow.cs]: adjacent*

or remove the TODO altogether

File: sdkproject/Assets/Mapbox/Unity/Editor/MapManagerEditor.cs [hunk 2]
```diff
@@ -158,6 +181,36 @@ public override void OnInspectorGUI()
...
+			{
+				var vectorDataProperty = serializedObject.FindProperty("_vectorData");
+
+
+				var layerProperty = vectorDataProperty.FindPropertyRelative("_layerProperty");
+				var layerSourceProperty = layerProperty.FindPropertyRelative("sourceOptions");
+				var sourceType = layerProperty.FindPropertyRelative("_sourceType");
+				VectorSourceType sourceTypeValue = (VectorSourceType)sourceType.enumValueIndex;
+				string streets_v7 = MapboxDefaultVector.GetParameters(VectorSourceType.MapboxStreets).Id;
+				//var layerSourceId = vectorDataProperty.FindPropertyRelative("Id");
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Editor/MapManagerEditor.cs]: Should we just remove some of these commented-out lines?

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/CoreVectorLayerPropertiesDrawer.cs [hunk 3]
```diff
@@ -15,8 +15,8 @@ public override void OnGUI(Rect position, SerializedProperty property, GUIConten
 			position.height = lineHeight;
 
 			// Draw label.
-			EditorGUI.PropertyField(new Rect(position.x, position.y, position.width, lineHeight), property.FindPropertyRelative("isActive"));
-			position.y += lineHeight;
+			//EditorGUI.PropertyField(new Rect(position.x, position.y, position.width, lineHeight), property.FindPropertyRelative("isActive"));
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/CoreVectorLayerPropertiesDrawer.cs]: Remove?

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/CoreVectorLayerPropertiesDrawer.cs [hunk 4]
```diff
@@ -47,9 +47,9 @@ public override float GetPropertyHeight(SerializedProperty property, GUIContent
 			var sourceTypeProperty = property.FindPropertyRelative("geometryType");
 
 			float height = 0.0f;
-			height += (((((VectorPrimitiveType)sourceTypeProperty.enumValueIndex == VectorPrimitiveType.Line)) ? 6.0f : 5.0f) * EditorGUIUtility.singleLineHeight);
+			height += (((((VectorPrimitiveType)sourceTypeProperty.enumValueIndex == VectorPrimitiveType.Line)) ? 5.0f : 4.0f) * EditorGUIUtility.singleLineHeight);
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/CoreVectorLayerPropertiesDrawer.cs]: Where do these float values come from? Can we make this less opaque?

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/LocationPrefabsLayerPropertiesDrawer.cs [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+			if (GUILayout.Button(new GUIContent("Add Layer"), (GUIStyle)"minibuttonleft"))
+			{
+
+				GUILayout.Space(EditorGUIUtility.singleLineHeight);
+
+				selectedLayers = layerTreeView.GetSelection();
+
+				prefabItemArray.arraySize++;
+				//subLayerArray.InsertArrayElementAtIndex(subLayerArray.arraySize);
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/LocationPrefabsLayerPropertiesDrawer.cs]: Remove?

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+				//include a remove button in the row
+				if (GUILayout.Button(new GUIContent(" X "), (GUIStyle)"minibuttonright", GUILayout.Width(30)))
+				{
+					coordinateProperties.DeleteArrayElementAtIndex(i);
+				}
+				EditorGUILayout.EndHorizontal();
+			}
+
+			EditorGUILayout.BeginHorizontal();
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs]: A lot of the GUI layout code seems a little messy/opaque (unnecessary newlines, num values that seem arbitrary, etc). It's ultimately a purely cosmetic change, but is there a plan at some point to comb through the code and clean it up?

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+		{
+			//Density slider
+			var densityProp = property.FindPropertyRelative("density");
+			if (Application.isPlaying)
+			{
+				GUI.enabled = false;
+			}
+
+			EditorGUILayout.PropertyField(densityProp, densitySlider);
+			var integ = densityProp.intValue;
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs]: Does "integ" get used? If so, can we rename it to something more meaningful?

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/SpawnPrefabOptionsDrawer.cs [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+		private GUIContent prefabContent = new GUIContent
+		{
+			text = "Prefab",
+			tooltip = "The prefab to be spawned"
+		};
+
+		private GUIContent scalePrefabContent = new GUIContent
+		{
+			text = "Scale down with world",
+			tooltip = "Turning on this option causes the prefab to scale down when you scale the map object"
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/SpawnPrefabOptionsDrawer.cs]: -> "The prefab will scale with the map object" [keep it short; feel free to tweak this if it's not exactly right]

File: sdkproject/Assets/Mapbox/Unity/LayerProperties/PrefabItemOptions.cs [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+		/// <summary>
+		/// Spawn at any location containing this name string
+		/// </summary>
+		public string nameString = "Name";
+
+		/// <summary>
+		/// The number of prefabs to spawn per-tile. Try a lower number if the map is cluttered.
+		/// </summary>
+		public Popularity popularity;
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/LayerProperties/PrefabItemOptions.cs]: Elsewhere I believe we call this density; any reason to use both names?

File: sdkproject/Assets/Mapbox/Unity/Map/AbstractMap.cs [hunk 10]
```diff
@@ -661,6 +662,173 @@ public void SetLoadingTexture(Texture2D loadingTexture)
...
+				_vectorData.AddVectorLayer(item);
+			}
+		}
+
+		/// <summary>
+		/// Places a game object prefabs all locations specified by the LatLon array.
+		/// </summary>
+		/// <param name="prefab"> A Game Object Prefab.</param>
+		/// <param name="LatLon">A Vector2d(Latitude Longitude) object</param>
+		public void SpawnPrefabAtGeoLocation(GameObject prefab, Vector2d[] LatLon, Action<List<GameObject>> callback = null, bool scaleDownWithWorld = true, string locationItemName = "New Location")
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/Map/AbstractMap.cs]: These SpawnPrefabAt functions have a lot in common; any way we can reduce redundancy?

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/LocationPrefabsLayerVisualizer.cs [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+			item.filterOptions.combinerType = item._combinerType;
+
+		}
+
+		/// <summary>
+		/// Gets the list of categories selected through the dropdown
+		/// </summary>
+		/// <returns>The selected categories list.</returns>
+		/// <param name="cat">Cat.</param>
+		private List<LocationPrefabCategories> GetSelectedCategoriesList(LocationPrefabCategories cat)
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/LocationPrefabsLayerVisualizer.cs]: Can we come up with a more descriptive (but unfortunately less cute) name than "cat"?

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/LocationPrefabsLayerVisualizer.cs [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+					//UnityTile space
+					latLonPoint.Add(new Vector3((float)(point.x / _extent * _rect.Size.x - (_rect.Size.x / 2)) * tile.TileScale, 
+					                            0, 
+					                            (float)((point.z) / _extent * _rect.Size.y - (_rect.Size.y / 2)) * tile.TileScale
+					                           ));
+
+					//add coordinate feature to feature points
+					feature.Points.Add(latLonPoint);
+
+					//HACK: silence errors when modifiers access feature.Data
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/LocationPrefabsLayerVisualizer.cs]: Is this hack still necessary, and if so, can we at least remove the comment that says it's a hack?

File: documentation/docs/05-changelog.md [hunk 13]
```diff
@@ -3,6 +3,7 @@
 *??/??/2018*
 
 ##### New Features
+- Add `Location Prefabs` section to `AbstractMap`. Enables users to spawn prefabs at points-of-interest on the map directly from the `AbstractMap` UI.
 - Add `OnUpdated` event to `AbstractMap`. Enables subscribers to get a notification when the map location and/or zoom gets updated.
```
Comment (atripathi-mb) [inline documentation/docs/05-changelog.md]: Please make the changelog more detailed. Add a list of major features that have been added.

File: sdkproject/Assets/Mapbox/Unity/Editor/MapManagerEditor.cs (hunk 2 above)
Comment (atripathi-mb) [reply sdkproject/Assets/Mapbox/Unity/Editor/MapManagerEditor.cs]: Agree with @morgane - Remove commented out lines.

File: sdkproject/Assets/Mapbox/Unity/Editor/MapManagerEditor.cs [hunk 14]
```diff
@@ -170,10 +223,10 @@ public override void OnInspectorGUI()
 
 		void ShowSection(SerializedProperty property, string propertyName)
 		{
-			EditorGUILayout.Space();
+    		EditorGUILayout.Space();
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/Editor/MapManagerEditor.cs]: Fix formatting.

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/CoreVectorLayerPropertiesDrawer.cs (hunk 4 above)
Comment (atripathi-mb) [reply sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/CoreVectorLayerPropertiesDrawer.cs]: These are the number of rows required for the UI based on the primitive. We can add a variable or code comment

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+			EditorGUILayout.EndHorizontal();
+
+			ShowDensitySlider(property);
+		}
+
+		private void ShowAddressOrLatLonUI(SerializedProperty property)
+		{
+
+
+			EditorGUILayout.BeginVertical();
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs]: Remove new lines.

File: sdkproject/Assets/Mapbox/Unity/Map/AbstractMap.cs [hunk 16]
```diff
@@ -661,6 +662,173 @@ public void SetLoadingTexture(Texture2D loadingTexture)
...
+		#region Location Prefabs Methods
+
+		/// <summary>
+		/// Places a game object prefabs at the specified LatLon on the Map.
+		/// </summary>
+		/// <param name="prefab"> A Game Object Prefab.</param>
+		/// <param name="LatLon">A Vector2d(Latitude Longitude) object</param>
+		public void SpawnPrefabAtGeoLocation(GameObject prefab, Vector2d LatLon, Action<List<GameObject>> callback = null, bool scaleDownWithWorld = true, string locationItemName = "New Location")
+		{
+			var coordinateArray = new string[] { LatLon.x + ", " + LatLon.y };
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/Map/AbstractMap.cs]: This method body can be simplified to `SpawnPrefabAtGeoLocation(prefab, coordinateArray,....)`

File: sdkproject/Assets/Mapbox/Unity/Map/AbstractMap.cs [hunk 17]
```diff
@@ -661,6 +662,173 @@ public void SetLoadingTexture(Texture2D loadingTexture)
...
+				item.OnAllPrefabsInstantiated += callback;
+			
+			if (!_vectorData.LayerProperty.vectorSubLayers.Contains(item))
+			{
+				_vectorData.LocationPrefabsLayerProperties.locationPrefabList.Add(item);
+				_vectorData.AddVectorLayer(item);
+			}
+		}
+
+  		#endregion
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/Map/AbstractMap.cs]: Can we refactor the redundant code in all 4 prefab functions ? Only difference I see is setting up the prefabItem options.

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/LocationPrefabsLayerVisualizer.cs [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+	{
+		private int maxDensity = 30; //This value is same as the density's max range value in PrefabItemOptions
+
+		public void SetProperties(PrefabItemOptions item, LayerPerformanceOptions performanceOptions)
+		{
+			SubLayerProperties = item;
+			_performanceOptions = performanceOptions;
+
+			if (!item.isActive)
+				return;
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/LocationPrefabsLayerVisualizer.cs]: Shouldn't the properties be set regardless of the "Active" status of the layer ?

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs [hunk 19]
```diff
@@ -14,26 +14,52 @@
...
 
-		ModifierStackBase _defaultStack;
+		protected ModifierStackBase _defaultStack;
 		private HashSet<ulong> _activeIds;
 		private Dictionary<UnityTile, List<ulong>> _idPool; //necessary to keep _activeIds list up to date when unloading tiles
 
 		private string _key;
 
+
+
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs]: Remove new lines?

File: sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs [hunk 20]
```diff
@@ -55,6 +81,7 @@ public void SetProperties(VectorSubLayerProperties properties, LayerPerformanceO
 			switch (properties.coreOptions.geometryType)
 			{
 				case VectorPrimitiveType.Point:
+					break;
```
Comment (atripathi-mb) [inline sdkproject/Assets/Mapbox/Unity/MeshGeneration/LayerVisualizers/VectorLayerVisualizer.cs]: This break is not correct, snapterrain modifier is required for both custom & point primitive types.

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+	using UnityEditor;
+	using System;
+	using System.Collections.Generic;
+
+	[CustomPropertyDrawer(typeof(PrefabItemOptions))]
+	public class PrefabItemOptionsDrawer : PropertyDrawer
+	{
+
+		static float _lineHeight = EditorGUIUtility.singleLineHeight;
+		private int shiftLeftPixels = -22;
```
[the remark this answers is not in the document]
Comment (jim-martin) [reply sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs]: This is actually difficult to address cleanly. The purpose is to move the input field of indented custom items to align with unity's generic ones:

With:
screenshot 2018-05-01 12 02 27


Without:
screenshot 2018-05-01 11 58 45

I wasn't able to find a way to get the actual indent value to offset these fields correctly. The closest I can find is `EditorGUI.IndentRect(position).width` which doesn't produce the right offset.

File: sdkproject/Assets/Mapbox/Unity/Map/AbstractMap.cs (hunk 16 above)
Comment (jim-martin) [reply sdkproject/Assets/Mapbox/Unity/Map/AbstractMap.cs]: I think it's better to take a Vector2d in the method parameters and handle the string conversion. 

We're limited to strings in the UI and sublayer properties to keep things compatible with the `GeoCodeAttributeSearchWindow` for now, but in script it's easier to work with lat/lon as a vector2d than a comma separated string.

File: sdkproject/Assets/Mapbox/Examples/6_LocationPrefabs/LocationPrefabsScriptExample.cs [hunk 22]
```diff
@@ -0,0 +1,74 @@
+﻿                                                                                             namespace Mapbox.Examples
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Examples/6_LocationPrefabs/LocationPrefabsScriptExample.cs]: Is this line really so heavily indented?

File: sdkproject/Assets/Mapbox/Examples/6_LocationPrefabs/LocationPrefabsScriptExample.cs [hunk 23]
```diff
@@ -0,0 +1,74 @@
...
+	using Mapbox.Utils;
+
+	public class LocationPrefabsScriptExample : MonoBehaviour
+	{
+
+		public AbstractMap map;
+
+		//prefab to spawn
+		public GameObject beforeInitialize;
+		//cache of spawned gameobjects
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Examples/6_LocationPrefabs/LocationPrefabsScriptExample.cs]: Let's stick to calling them prefabs for consistency

File: sdkproject/Assets/Mapbox/Examples/6_LocationPrefabs/LocationPrefabsScriptExample.cs [hunk 24]
```diff
@@ -0,0 +1,74 @@
...
+		private List<GameObject> _beforeInitializeInstances;
+
+		public GameObject afterInitialize;
+		private List<GameObject> _afterInitializeInstances;
+
+		public GameObject afterLoaded;
+		private List<GameObject> _afterLoadedInstances;
+
+
+
```
Comment (morgane) [inline sdkproject/Assets/Mapbox/Examples/6_LocationPrefabs/LocationPrefabsScriptExample.cs]: Are all these newlines necessary?

File: sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs (hunk 6 above)
Comment (morgane) [reply sdkproject/Assets/Mapbox/Unity/Editor/PropertyDrawers/PrefabItemOptionsDrawer.cs]: I like the sound of that

**Pull request merged**